sandbox/cgroup/freezer.go

```diff
@@ -82,36 +86,44 @@ var ThawSnapProcesses = thawSnapProcessesImplV1
 // freezeSnapProcessesImplV1 freezes all the processes originating from the given snap.
 // Processes are frozen regardless of which particular snap application they
 // originate from.
-func freezeSnapProcessesImplV1(snapName string) error {
+var freezeSnapProcessesImplV1 = func(ctx context.Context, snapName string) error {
 	fname := filepath.Join(freezerCgroupV1Dir, fmt.Sprintf("snap.%s", snapName), "freezer.state")
-	if err := os.WriteFile(fname, []byte("FROZEN"), 0644); err != nil && os.IsNotExist(err) {
+	if err := os.WriteFile(fname, []byte("FROZEN"), 0644); err != nil && errors.Is(err, fs.ErrNotExist) {
 		// When there's no freezer cgroup we don't have to freeze anything.
 		// This can happen when no process belonging to a given snap has been
 		// started yet.
 		return nil
 	} else if err != nil {
-		return fmt.Errorf("cannot freeze processes of snap %q, %v", snapName, err)
+		return fmt.Errorf("cannot freeze processes of snap %q, %w", snapName, err)
 	}
-	for i := 0; i < 30; i++ {
+
+	// Add an upper bound to the timeout if cgroup is stuck at freeze state
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, maxFreezeTimeout)
+	defer cancel()
+	ticker := time.NewTicker(100 * time.Millisecond)
```

Comment:
Please add `defer ticker.Stop()` with a note to perhaps remove this once we can depend on Go 1.23. See https://pkg.go.dev/time#NewTicker for rationale.